### PR TITLE
Added possibility of changing column length for existing fields

### PR DESCRIPTION
**Pull request by konadrian**
Added possibility of changing maxlength
ChangeMaxLength

### CODE REVIEW

**Comment by ghost** [comment]
## Black Duck Security Report

Merging #17204 into developer will not change security risk.

[Click here to see full report](https://copilot.blackducksoftware.com/github/repos/YetiForceCompany/YetiForceCRM/pull-requests/17204)

File: modules/Vtiger/uitypes/Base.php [hunk 1]
```diff
@@ -576,4 +576,26 @@ public function getFieldInfo(): array
 	{
 		return $this->getFieldModel()->loadFieldInfo();
 	}
+
+	public function isColumnLengthChangeAllowed(): bool
+	{
+		return true;
+	}
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/Base.php]: add property to class

File: modules/Vtiger/uitypes/Base.php [hunk 2]
```diff
@@ -576,4 +576,26 @@ public function getFieldInfo(): array
...
+
+	public function validateColumnLength($newColumnLength): bool
+	{
+		$minColumnLength = 0;
+		$maxColumnLength = 255;
+		$newColumnLength = (int) $newColumnLength;
+		if ($newColumnLength > $minColumnLength && $newColumnLength <= $maxColumnLength) {
+			return true;
+		}
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/Base.php]: ?

File: modules/Vtiger/uitypes/Base.php [hunk 3]
```diff
@@ -576,4 +576,26 @@ public function getFieldInfo(): array
...
+		throw new \App\Exceptions\Security('ERR_ILLEGAL_FIELD_VALUE||' . $this->getFieldModel()->getName() . '||' . $this->getFieldModel()->getModuleName() . '||' . $newColumnLength, 406);
+	}
+
+	public function hasColumnLengthChanged($newColumnLength)
+	{
+		$dbColumnStructure = $this->getFieldModel()->getDBColumnType(false);
+		return (int) $newColumnLength !== $dbColumnStructure['size'];
+	}
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/Base.php]: hmm..

File: public_html/layouts/basic/modules/Settings/LayoutEditor/resources/EditField.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
+/* {[The file is published on the basis of YetiForce Public License 5.0 that can be found in the following directory: licenses/LicenseEN.txt or yetiforce.com]} */
+'use strict';
+
+$.Class(
+	'Settings_LayoutEditor_EditField_JS',
```
Comment (rskrzypczak) [inline public_html/layouts/basic/modules/Settings/LayoutEditor/resources/EditField.js]: ?

File: modules/Vtiger/uitypes/Base.php (hunk 1 above)
Comment (konadrian) [reply modules/Vtiger/uitypes/Base.php]: Length method

File: modules/Vtiger/uitypes/Base.php (hunk 2 above)
Comment (konadrian) [reply modules/Vtiger/uitypes/Base.php]: Make sure that string is Base, create uitype fo string? Check system field for don't edit length. Return on end of method. True or false on return without throw

File: modules/Vtiger/uitypes/Base.php (hunk 3 above)
Comment (konadrian) [reply modules/Vtiger/uitypes/Base.php]: Try to change it,  App/Fixer is some verification of column length
hasColumnLengthChanged ???

File: public_html/layouts/basic/modules/Settings/LayoutEditor/resources/EditField.js (hunk 4 above)
Comment (konadrian) [reply public_html/layouts/basic/modules/Settings/LayoutEditor/resources/EditField.js]: move from layout editor

File: modules/Vtiger/uitypes/Base.php [hunk 5]
```diff
@@ -576,4 +576,38 @@ public function getFieldInfo(): array
...
 	}
+
+	/**
+	 * Method is designed to determine whether it is possible to change the length of a column in a database structure.
+	 *
+	 * @return bool
+	 */
+	public function isResizableColumn(): bool
+	{
+		return false;
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/Base.php]: move to property

File: modules/Settings/LayoutEditor/actions/Field.php [hunk 6]
```diff
@@ -324,4 +329,24 @@ public function createSystemField(App\Request $request): void
...
+	private function changeColumnLength(Settings_LayoutEditor_Field_Model $fieldInstance, string $columnLength)
+	{
+		$uitypeModel = $fieldInstance->getUITypeModel();
+		if ($uitypeModel->isResizableColumn() && $uitypeModel->isColumnLengthIncreased($columnLength) && $uitypeModel->validateColumnLength($columnLength)) {
+			$dbColumnStructure = $fieldInstance->getDBColumnType(false);
+			$columnType = $dbColumnStructure['type'];
+			$db = App\Db::getInstance();
+			$db->createCommand()->alterColumn($fieldInstance->getTableName(), $fieldInstance->getColumnName(), "{$columnType}({$columnLength})")->execute();
+			$fieldInstance->set('maximumlength', $fieldInstance->getRangeValues());
+		}
```
Comment (rskrzypczak) [inline modules/Settings/LayoutEditor/actions/Field.php]: You won't always need to change the column type. By default, only maximumlength is to be modified, and the column type in special cases.

File: modules/Vtiger/uitypes/Integer.php [hunk 7]
```diff
@@ -98,4 +101,11 @@ public function getQueryOperators()
 	{
 		return array_merge(['e', 'n', 'l', 'g', 'm', 'h', 'y', 'ny'], \App\Condition::FIELD_COMPARISON_OPERATORS);
 	}
+
+	/** {@inheritdoc} */
+	public function validateMaximumLength(int $newMinValue, int $newMaxValue): bool
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/Integer.php]: why the same code need to be set in difference uitype?

File: modules/Vtiger/uitypes/String.php [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	/** {@inheritdoc} */
+	protected $isResizableColumn = true;
+
+	/** {@inheritdoc} */
+	public function validateMaximumLength(int $minimumLength, int $maximumLength): bool
+	{
+		$allowedMaxValue = 255;
+		$newColumnLength = (int) $maximumLength;
+		return $newColumnLength <= $allowedMaxValue;
+	}
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/String.php]: you have something like this here: https://github.com/YetiForceCompany/YetiForceCRM/pull/17204/files#diff-70152fd4caea113f66f8f617348ab9eedf6d5f22bd6e7ea23a64d558211b6374R1713-R1716
and in the other uitype you use: https://github.com/YetiForceCompany/YetiForceCRM/pull/17204/files#diff-817852a30f3e1bf77315f8c697104e60cc94ad9a3b1448d2f231ea81e317cfe1R95-R99
why you can't use the same here?

**Review by rskrzypczak — CHANGES_REQUESTED** [review]

File: modules/Vtiger/uitypes/String.php [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+	/** {@inheritdoc} */
+	public function changeMaximumLength(int $minimumLength, int $maximumLength): void
+	{
+		if ($this->isResizableColumn() && $this->validateMaximumLength($minimumLength, $maximumLength)) {
+			$fieldInstance = $this->getFieldModel();
+			if ($this->isColumnLengthIncreased($maximumLength)) {
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/String.php]: Do you really need the "isColumnLengthIncreased" function?
you call the getDBColumnType(false) function anyway.

**Pull request merged**